From 4f8532b0edead154b1daffb42397483ad74ad79b Mon Sep 17 00:00:00 2001
From: Devin Lin <espidev@gmail.com>
Date: Thu, 15 Jan 2026 22:15:19 -0500
Subject: [PATCH] Fix lockscreen overlay behavior

- Ensure that lockscreen state is fetched at start
- Register with lockscreen overlay protocol before window is shown
  (required by protocol)
---
 plasma-dialer/src/lockscreenutils.cpp | 15 +++++++--------
 plasma-dialer/src/main.cpp            |  4 +++-
 plasma-dialer/src/qml/Main.qml        |  9 +++++----
 3 files changed, 15 insertions(+), 13 deletions(-)

diff --git a/plasma-dialer/src/lockscreenutils.cpp b/plasma-dialer/src/lockscreenutils.cpp
index 8b41b910..b197992f 100644
--- a/plasma-dialer/src/lockscreenutils.cpp
+++ b/plasma-dialer/src/lockscreenutils.cpp
@@ -31,7 +31,12 @@ LockScreenUtils::LockScreenUtils(QObject *parent)
                                                           QStringLiteral("org.freedesktop.ScreenSaver"),
                                                           QStringLiteral("GetActive"));
 
-    QDBusConnection::sessionBus().callWithCallback(request, this, SLOT(slotLockscreenActiveChanged(bool)), SLOT(dbusError(QDBusError)));
+    QDBusReply<bool> reply = QDBusConnection::sessionBus().call(request);
+    if (reply.isValid()) {
+        m_lockscreenActive = reply.value();
+    } else {
+        qDebug() << "Error fetching lockscreen state using DBus:" << reply.error().message();
+    }
 
     QDBusConnection::sessionBus().connect(QStringLiteral("org.freedesktop.ScreenSaver"),
                                           QStringLiteral("/ScreenSaver"),
@@ -61,13 +66,7 @@ void LockScreenUtils::slotLockscreenActiveChanged(bool active)
         m_lockscreenActive = active;
 
         Q_EMIT lockscreenActiveChanged();
-
-        // we don't want to trigger a lockscreen changing signal for the first property fetch (in constructor),
-        // since it's just getting the current state
-        if (m_firstPropertySet) {
-            m_lockscreenActive ? Q_EMIT lockscreenLocked() : Q_EMIT lockscreenUnlocked();
-        }
-        m_firstPropertySet = true;
+        m_lockscreenActive ? Q_EMIT lockscreenLocked() : Q_EMIT lockscreenUnlocked();
     }
 }
 
diff --git a/plasma-dialer/src/main.cpp b/plasma-dialer/src/main.cpp
index edb19f33..b9ac4ce2 100644
--- a/plasma-dialer/src/main.cpp
+++ b/plasma-dialer/src/main.cpp
@@ -174,6 +174,9 @@ int main(int argc, char **argv)
     QWindow *window = qobject_cast<QWindow *>(engine.rootObjects().at(0));
     Q_ASSERT(window);
 
+    // Register with lockscreen overlay protocol at startup before window is visible
+    allowAboveLockscreen(window);
+
     QObject::connect(&service, &KDBusService::activateRequested, window, [&window](const QStringList &arguments) {
         qDebug() << "Window activation requested over DBus";
 
@@ -185,7 +188,6 @@ int main(int argc, char **argv)
             }
         }
 
-        allowAboveLockscreen(window);
         raiseWindow(window);
     });
 
diff --git a/plasma-dialer/src/qml/Main.qml b/plasma-dialer/src/qml/Main.qml
index d17bc749..61881ca0 100644
--- a/plasma-dialer/src/qml/Main.qml
+++ b/plasma-dialer/src/qml/Main.qml
@@ -200,13 +200,14 @@ Kirigami.ApplicationWindow {
     }
 
     Connections {
-        function onDepthChanged() {
-            if (root.lockscreenMode && applicationWindow().pageStack.layers.depth < 2) {
-                console.log("Exiting since the call page was dismissed on the lockscreen");
+        function onLockscreenLocked() {
+            if (!ActiveCallModel.active) {
+                console.log("There is no call ongoing and we are in the lockscreen, exiting");
                 Qt.quit();
             }
         }
-        target: applicationWindow().pageStack.layers
+
+        target: LockScreenUtils
     }
 
     Connections {
-- 
GitLab

